### PR TITLE
Asynchronous transactions (polling based for now).

### PR DESCRIPTION
**Pull request by gavofyork**
- Add `eth_postTransaction` as an async alternative to `eth_sendTransaction`. Returns a Confirmation promise ID.
- Add `eth_checkTransaction`, takes a single argument - a Confirmation promise ID and returns
one of:
  - Transaction hash (signed and submitted).
  - `null` (still pending, call again later),
  - Zero hash (will never be signed).

### CODE REVIEW

**Comment by gavofyork** [comment]
Additional web3.js code required to make this work:

```javascript
web3._extend({
	property: 'eth',
	methods: [
		new web3._extend.Method({
			name: 'postTransaction',
			call: 'eth_postTransaction',
			params: 1,
			inputFormatter: [web3._extend.formatters.inputCallFormatter]
		})
	]
});

web3._extend({
	property: 'eth',
	methods: [
		new web3._extend.Method({
			name: 'checkTransaction',
			call: 'eth_checkTransaction',
			params: 1
		})
	]
});

{
	var postTransaction = web3.eth.postTransaction.bind(web3.eth);
	var sendTransaction = web3.eth.sendTransaction.bind(web3.eth);
	web3.eth.sendTransaction = function(options, f) {
		// No callback - use compatibility sync API.
		if (typeof f != "function")
			return sendTransaction(options);
		// Callback - use async API.
		var id = postTransaction(options);
		var timer_id = window.setInterval(check, 500);
		function check() {
			let r = web3.eth.checkTransaction(id);
			if (typeof r == 'string') {
				clearInterval(timer_id);
				if (r == "0x0000000000000000000000000000000000000000000000000000000000000000")
					f("Rejected", r);
				else
					f(null, r);
			}
		}
	}
}
```

File: rpc/src/v1/helpers/signing_queue.rs [hunk 1]
```diff
@@ -125,34 +126,43 @@ impl ConfirmationToken {
...
+			// Check the result...
+			match *self.result.lock() {
+				// Waiting and deadline not yet passed continue looping.  
+				ConfirmationResult::Waiting if now < deadline => {}
+				// Anything else - return.
+				ref a => return a.clone(),
 			}
-			// Park thread (may wake up spuriously)
+			// wait a while longer - maybe the solution will arrive.
 			thread::park_timeout(deadline - now);
```
Comment (tomusdrw) [inline rpc/src/v1/helpers/signing_queue.rs]: Can panic when `now > deadline`

File: rpc/src/v1/impls/eth_signing.rs [hunk 2]
```diff
@@ -59,6 +61,7 @@ impl<C, M> EthSigningQueueClient<C, M> where C: MiningBlockChainClient, M: Miner
 			accounts: Arc::downgrade(accounts),
 			client: Arc::downgrade(client),
 			miner: Arc::downgrade(miner),
+			pending: Mutex::new(HashMap::new()),
```
Comment (tomusdrw) [inline rpc/src/v1/impls/eth_signing.rs]: It can leak, maybe `TransientHashMap` (the same one is used for filters) could be used here?

**Comment by tomusdrw** [comment]
JS code is invalid in general case - The code should be applied only if signer is enabled.

Minor things: 
1. `Rejected` -> `{ message: 'rejected' }` for compatibility with Mist.
2. If `postTransaction` or `checkTransaction` fail (for instance when node is turned off or API returns `invalid params error`) `callback` is not invoked (you need to try/catch the invocations when using sync methods)

**Comment by gavofyork** [comment]
signer should always be used with UI and in any case JS code is not part of this PR.

**Pull request merged**